### PR TITLE
Add queue overflow handler in asyncsender.

### PR DESCRIPTION
**Pull request by pguz**
Hi,

In order to guarantee that the logging process won't block under any circumstance the app, `queue_circular` flag can be used. In that case it can cause loss of events if the queue fills up completely. 

However sometimes there is a need to handle such cases in some way. That's why I propose adding `queue_overflow_handler` parameter to `asyncsender`.

Please consider my pull request.

Cheers!

### CODE REVIEW

**Comment by arcivanov** [comment]
Please add a test to cover.

File: fluent/asyncsender.py [hunk 1]
```diff
@@ -132,5 +135,13 @@ def _send_loop(self):
         finally:
             self._close()
 
+    def _call_queue_overflow_handler(self, discarded_bytes):
+        try:
+            if self._queue_overflow_handler:
```
Comment (arcivanov) [inline fluent/asyncsender.py]: I don't like checks like that in the critical paths, as it's going to be repeated every call.
Can we push it up into the constructor like we do here https://github.com/fluent/fluent-logger-python/blob/master/fluent/handler.py#L86?

**Comment by pguz** [comment]
I have just added improved test.

**Comment by arcivanov** [comment]
I'm travelling and will try to get to this tonight.

**Comment by pguz** [comment]
up

**Comment by arcivanov** [comment]
@pguz sorry, it was Pending review for two weeks now, because I always forget to actually submit the review :(

File: fluent/asyncsender.py [hunk 2]
```diff
@@ -109,7 +114,8 @@ def _send(self, bytes_):
             if self._queue_circular and self._queue.full():
                 # discard oldest
                 try:
-                    self._queue.get(block=False)
+                    discarded_bytes = self._queue.get(block=False)
+                    self._queue_overflow_handler(discarded_bytes)
```
[the remark this answers is not in the document]
Comment (pguz) [reply fluent/asyncsender.py]: I have just resolved it. I hope it does a job.

**Comment by pguz** [comment]
Locally I am not receiving these errors. Is it testing environment issue?

**Comment by arcivanov** [comment]
Nope, that's a dependency issue, i.e. your local testing environment issue :)

**Comment by jgoclawski** [comment]
Hi @arcivanov, this PR looks like a nice addition. Would you have any hints as to why one test fails only o Python 2.7 on Travis? What can be done to move this forward?

```
FAIL: test_simple (tests.test_asynchandler.TestHandlerWithCircularQueueHandler)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/fluent/fluent-logger-python/tests/test_asynchandler.py", line 371, in test_simple
    log.info({'cnt': 2, 'from': 'userA', 'to': 'userB'})
AssertionError: QueueOverflowException not raised
```

**Comment by pguz** [comment]
Hi, I fixed this issue writing more `defensive` test. `QueueOverflowException not raised` error was not caused by python2.7 itself, but rather multithreading issues. We cannot be sure that QueueOverflowException is always thrown, that's why in the last commit I assumed that this exception may not be thrown, however I wrapped in print statement number of times this exception is excepted to be thrown. 

`print('Exception raised: {} (expected 3)'.format(exc_counter))`

We can see that in newer series of tests for python 2.7 tests are passing and even the condition which was not passed last time is passing now. However this time the problematic condition would not pass `pypy3.6-7.1.1` tests. However due to defensive nature of tests `pypy3.6-7.1.1 tests` are passing. I am sure that running tests one more time for `pypy3.6-7.1.1 tests` would pass the problematic condition.

**Comment by pguz** [comment]
up

**Pull request merged**